### PR TITLE
feat: add rich-indexer which is another built-in indexer based on relational database

### PR DESCRIPTION
**Pull request by EthanYuan**
### What problem does this PR solve?

Problem Summary:

CKB already has a built-in indexer, which is based on a key-value database for indexing. Rich-Indexer is a complement to the original indexer, as it is based on a relational database, providing support for more complex and flexible queries.

Currently, Rich-Indexer supports two types of database drivers. 

1. One is the embedded SQLite, which users can start using without any configuration.
2. The other is PostgreSQL, which can undergo independent, customized configurations for both software and hardware. Additionally, it allows users to engage in secondary development based on it.

### What is changed and how it works?

What's Changed:

Once the Rich-Indexer is activated, the CKB node's RPC based on the Rich-Indexer will gain additional capabilities.

| INDEXER RPC          | Indexer           | Rich-Indexer       |
|---------------------|-------------------|--------------------|
| `get_cells` script args `partial` mode search     | ❌     | ✔️      |
| `get_cells` cell data filter(`prefix\|exact\|partial`)     | ✔️     | ✔️      |
| `get_transactions` script args `partial` mode search | ❌          | ✔️   |
| `get_transactions` cell data filter(`prefix\|exact\|partial`)     | ❌         |  ✔️      |
| `get_cells_capacity` script args `partial` mode search        |    ❌    |    ✔️     |
| `get_cells_capacity` cell data filter(`prefix\|exact\|partial`)        |    ✔️    |   ✔️     |

In addition, Rich-Indexer can also expose full block data in the form of a PostgreSQL database, providing the most basic and flexible indexing format accessible via SQL and compatible with any programming language.

Please refer to the [README](https://github.com/EthanYuan/ckb/tree/indexer-r/util/rich-indexer#readme) for more detailed information

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code ci-runs-only: [ quick_checks,linters ]

Side effects

- Performance regression

### Release note 

```release-note
Title Only: Include only the PR title in the release note.
```

### CODE REVIEW

File: resource/ckb.toml [hunk 1]
```diff
@@ -199,3 +199,20 @@ block_uncles_cache_size    = 30
 # [indexer_v2]
 # # Indexing the pending txs in the ckb tx-pool
 # index_tx_pool = false
+# block_filter = "block.header.number.to_uint() > \"10000000\".to_uint()"
+# cell_filter = "let script = output.type;script.code_hash == \"0xbbad126377d45f90a8ee120da988a2d7332c78ba8fd679aab478a19d6c133494\" && script.hash_type == \"data1\""
```
Comment (chenyukang) [inline resource/ckb.toml:203]: we should remove this?

File: rpc/src/service_builder.rs [hunk 2]
```diff
@@ -201,15 +204,51 @@ impl<'a> ServiceBuilder<'a> {
...
+        let indexer_r = IndexerRService::new(
+            ckb_secondary_db,
+            pool_service.clone(),
+            indexer_config,
+            shared.async_handle().clone(),
+        );
+        let indexer_r_handle = indexer_r.handle();
+        let indexer_r_rpc_methods = IndexerRRpcImpl::new(indexer_r_handle).to_delegate();
+
+        if self.config.indexer_r_enable() {
```
Comment (chenyukang) [inline rpc/src/service_builder.rs:242]: `indexer_r` seems too short to distinguish from `indexer`, maybe `db_indexer`?

File: resource/ckb.toml [hunk 3]
```diff
@@ -115,7 +115,7 @@ listen_address = "127.0.0.1:8114" # {{
 # Default is 10MiB = 10 * 1024 * 1024
 max_request_body_size = 10485760
 
-# List of API modules: ["Net", "Pool", "Miner", "Chain", "Stats", "Subscription", "Experiment", "Debug", "Indexer"]
+# List of API modules: ["Net", "Pool", "Miner", "Chain", "Stats", "Subscription", "Experiment", "Debug", "Indexer", "IndexerR"]
```
Comment (eval-exec) [inline resource/ckb.toml:118]: How about `"RelationalIndexer"`

File: util/indexer-r/src/indexer/mod.rs [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+                }
+            }
+        }
+        bulk_insert_output_table(&output_cell_rows, tx).await?;
+        bulk_insert_script_table(&script_set, tx).await?;
+
+        // The output needs to be inserted into the db-transaction before the input traversal.
+        // This is to cope with the case where the output is spent in a transaction in the same block,
+        // because the input needs to query the corresponding output cell when doing a cell filter.
+        for (tx_index, tx_view) in tx_views.iter().enumerate() {
```
Comment (eval-exec) [inline util/indexer-r/src/indexer/mod.rs:221]: How about use `.skip(1)` to skip cellbase tx?

Comment (EthanYuan) [reply util/indexer-r/src/indexer/mod.rs:221]: yes, it looks cleaner with skip. I'll change it to something like this:

```rust
// Skip the first cellbase transaction.
for tx_view in tx_views.iter().skip(1) {
...
}
```

File: rpc/src/service_builder.rs (hunk 2 above)
Comment (EthanYuan) [reply rpc/src/service_builder.rs:231]: Renamed rich indexer.

File: resource/ckb.toml (hunk 1 above)
Comment (EthanYuan) [reply resource/ckb.toml:203]: Keeping these demonstrates the available features.

File: util/rich-indexer/README.md
Comment (15168316096) [inline util/rich-indexer/README.md:1]: indexer rpc need update:  get_cells_capaciry -> get_cells_capacity

**Review by 15168316096 — COMMENTED** [review]
https://github.com/nervosnetwork/ckb/pull/4224/files#diff-c2f7597ec2fcd27837aa4a61160b41fc2bdbd9bc5afba75cf8778b9a2ff6c04fR53-R54

**Comment by 15168316096** [comment]
@EthanYuan 
`# API module list: ["Net", "Pool", "Miner", "Chain", "Stats", "Subscription", "Experiment", "Debug", "Indexer", "RichIndexer"]
module = ["Net", "Pool", "Miner", "Chain", "Stats", "Subscription", "Experiment", "Indexer", "RichIndexer"]`
Using the startup command under this configuration will report an error.
./ckb run --indexer
Configuration error: Indexer and rich indexer cannot be set at the same time
This configuration only opens the api module. I think this detection error is redundant?

**Comment by eval-exec** [comment]
https://github.com/cryptape/acceptance-internal/issues/698#issuecomment-1985054908

**Review by zhangsoledad — APPROVED** [review]

**Pull request merged**